cEP-0015.md

```diff
@@ -0,0 +1,153 @@
+# corobo Enhancement (security, tests and configurability)
+
+| Metadata |                                                          |
+| -------- | -------------------------------------------------------- |
+| cEP      | 15                                                       |
+| Version  | 1.0                                                      |
+| Title    | corobo Enhancement (security, tests and configurability) |
+| Authors  | Nitanshu Vashistha <mailto:nitanshu.vzard@gmail.com>     |
+| Status   | Proposed                                                 |
+| Type     | Process                                                  |
+
+## Abstract
+
+This cEP describes the details of enhancement of
+[corobo](https://github.com/coala/corobo) in terms of security, tests,
+configurability and the new plugins that are to be added to corobo as part
+of the
+[GSoC Project](https://summerofcode.withgoogle.com/projects/#6603667076022272).
+
+## Security
+
+Security has been one of the major concerns due to some past experiences.
+These are the changes that will be implemented to the existing plugins
+in order to harden the security.
+
+1. Give developers the right to invite newcomers to the organization.
+   This will involve changing the bot command to `corobo invite <@username>`.
+
+2. Remove the automatic invite due to `Hello World` message.
+
+3. Remove `invite me` bot command.
+
+4. Stop newcomers from getting assigned to more than one newcomer issue.
+
+5. Make all LabHub plugins require being a member of the organization. This
+   will be done by creating a `cmdfilter` to check if the user is member of
+   team or not.
+
+   ```
+   @cmdfilter
+   def check_membership(self, msg, cmd, args, dry_run)
+        usr = msg.frm.nik
+        teams = [self.TEAMS[self.GH_ORG_NAME + 'newcomers'],
+                 self.TEAMS[self.GH_ORG_NAME + 'developers'],
+                 self.TEAMS[self.GH_ORG_NAME + 'maintainers']
+                ]
+        for team in teams:
+            if team.is_member(usr):
+                return msg, cmd, args
+
+        return None, None, None
+   ```
+
+6. Add callback message to warn users who are spamming the room.
+
+7. Add bot command to ban a user from all rooms at once. The goal of
+   this feature is to prevent spamming as someone with many accounts
+   can't be stopped to join a room.
+
+## Tests
+
+### Why improve existing tests and infrastructure?
+
+The default TestBase provided by errbot is not enough for testing plugins like
+LabHub, which required intensive mocking. One of the examples of this case is
+issues like [this](https://github.com/errbotio/errbot/issues/1086).
+
+### How to improve existing tests and infrastructure?
+
+This will involve making changes upstream in
+[Errbot](https://github.com/errbotio/errbot/) and extend the existing testing
+infrastructure to implement better testing for plugins like LabHub.
```

Comment:
We started to brainstorm about it with @nvzard. We are thinking about extending a testbot with a feature to be able to inject mocks in the plugins.
This would look something like this (but needs to be tried out to be sure it is user friendly):
```python
import pytest

fakeObj1 = FakeObj()  # your mock
extra_plugin_dir = ...  # path to your plugin to test (already implemented)
inject_in_plugins = { 'PluginName': {'field_name1': fakeObj1,
                                                         'field_name2': fakeObj2}}
def test_mycommand(testbot):
    assert 'something' in testbot.exec_command('!my_command')
```
Another way or complementary way might be (if we want to change mocks at every test)
```python
def test_mycommand(testbot):
    testbot.inject_mock('PluginName', 'field_name1', fakeObj1)
    assert 'something' in testbot.exec_command('!my_command')
```

Reply:
This looks good :D :+1:

Reply:
I suggest it would be better if the `inject_mock` takes a dict of mock objects as parameter  and the user is able to inject the dict for every command_test:
```
def test_mycommand(testbot):
    mock_config = {'PluginName': {'field_name1': fakeObj1,
                                   'field_name2': fakeObj2}}
    testbot.config_mocking(mock_config)
    assert 'something' in testbot.exec_command('!my_command')
```